napari/viewer.py

```diff
@@ -56,6 +64,28 @@ def __init__(
 
         self._window = Window(self, show=show)
         self._instances.add(self)
+        if console_local_update:
+            self._caller_frame_locals = self._find_calling_frame().f_locals
+        else:
+            self._caller_frame_locals = {}
+
+    def _find_calling_frame(self):
+        frame = inspect.currentframe()
+
+        while frame:
+
+            name = frame.f_globals.get("__name__", None)
+            if name in ('__main__', '<run_path>'):
+                return frame
+            if (
+                name != 'napari'
+                and name != '<run_path>'
+                and 'runpy' not in name
+                and not name.startswith('napari.')
+            ):
+                return frame
+
+            frame = frame.f_back
```

Comment:
This code reminded me of [`magic_name`](https://github.com/napari/napari/blob/842008f6df909bcbc391789badd7f2414eed6ec4/napari/utils/naming.py#L49-L100), which has a bunch of warnings about keeping a reference to the frame. We should combine the code paths to make sure that any frame-traversing bugs are contained to one place in the codebase.

Reply:
+1 that would be good to look into, or at least if we could move this code into a util function that could be independently tested